File: app/src/main/java/lu/kremi151/a3dtictactoe/view/GameBoard.java
/**
 * 3D Tic Tac Toe for Android
 * Copyright (C) 2018  Michel Kremer (kremi151)
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/

package lu.kremi151.a3dtictactoe.view;

import android.content.Context;
import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.Paint;
import android.util.AttributeSet;
import android.view.MotionEvent;
import android.view.View;

import lu.kremi151.a3dtictactoe.interfaces.FieldColorInterceptor;
import lu.kremi151.a3dtictactoe.interfaces.OnBoardTapListener;
import lu.kremi151.a3dtictactoe.util.GameCube;

public class GameBoard extends View {

    private Context mContext;
    private Paint fieldPaint, fieldContentPaint, fieldValuePaint;

    private boolean portrait;
    private int mViewWidth, mViewHeight;
    private float layerSize, layerSpacing, fieldSize;
    private OnBoardTapListener listener;
    private FieldColorInterceptor fieldColorInterceptor, valueColorInterceptor;

    private GameCube cube = new GameCube();

    public GameBoard(Context context) {
        super(context);
        init(context);
    }

    public GameBoard(Context context, AttributeSet attrs) {
        super(context, attrs);
        init(context);
    }

    public GameBoard(Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        init(context);
    }

    public GameBoard(Context context, AttributeSet attrs, int defStyleAttr, int defStyleRes) {
        super(context, attrs, defStyleAttr, defStyleRes);
        init(context);
    }

    public GameBoard setCube(GameCube cube){
        this.cube = cube;
        return this;
    }

    private void init(Context context) {
        mContext = context;
        fieldPaint = new Paint();
        fieldPaint.setColor(Color.DKGRAY);
        fieldPaint.setStrokeWidth(1);
        fieldPaint.setStyle(Paint.Style.STROKE);
        fieldValuePaint = new Paint(fieldPaint);
        fieldValuePaint.setStrokeWidth(4);
        fieldContentPaint = new Paint();
        fieldContentPaint.setColor(Color.WHITE);
        fieldContentPaint.setStyle(Paint.Style.FILL);
    }

    public void pause() {}

    public void resume() {}

    public void setListener(OnBoardTapListener listener){
        this.listener = listener;
    }

    public void setFieldColorInterceptor(FieldColorInterceptor interceptor){
        this.fieldColorInterceptor = interceptor;
    }

    public void setValueColorInterceptor(FieldColorInterceptor interceptor){
        this.valueColorInterceptor = interceptor;
    }

    @Override
    protected void onDraw(Canvas canvas) {
        canvas.save();
        canvas.drawColor(Color.WHITE);
        canvas.translate(1f, 1f);

        for(int z = 0 ; z < this.cube.depth() ; z++){
            for(int x = 0 ; x < this.cube.width() ; x++){
                for(int y = 0 ; y < this.cube.height() ; y++){
                    float left, top;
                    if(portrait){
                        left = (z * layerSpacing) + (x * fieldSize);
                        top = (z * layerSize) + (y * fieldSize);
                    }else{
                        left = (z * layerSize) + (x * fieldSize);
                        top = (z * layerSpacing) + (y * fieldSize);
                    }
                    if(this.fieldColorInterceptor != null){
                        fieldContentPaint.setColor(this.fieldColorInterceptor.getFieldColor(x, y, z, Color.WHITE));
                        canvas.drawRect(
                                left,
                                top,
                                left + fieldSize,
                                top + fieldSize,
                                fieldContentPaint);
                    }
                    canvas.drawRect(
                            left,
                            top,
                            left + fieldSize,
                            top + fieldSize,
                            fieldPaint);
                    if(this.valueColorInterceptor != null){
                        fieldValuePaint.setColor(valueColorInterceptor.getFieldColor(x, y, z, Color.BLACK));
                    }
                    switch (cube.valueAt(x, y, z)){
                        case CIRCLE:
                            canvas.drawCircle(left + (fieldSize / 2), top + (fieldSize / 2), fieldSize * 0.4f, fieldValuePaint);
                            break;
                        case CROSS:
                            drawCross(canvas, left + (fieldSize * 0.1f), top + (fieldSize * 0.1f), left + (fieldSize * 0.9f), top + (fieldSize * 0.9f), fieldValuePaint);
                            break;
                    }
                }
            }
        }

        canvas.restore();
    }

    private void drawCross(Canvas canvas, float left, float top, float right, float bottom, Paint paint){
        canvas.drawLine(left, top, right, bottom, paint);
        canvas.drawLine(left, bottom, right, top, paint);
    }

    @Override
    protected void onSizeChanged(int w, int h, int oldw, int oldh) {
        super.onSizeChanged(w, h, oldw, oldh);
        w -= 2;
        h -= 2;
        this.portrait = h >= w;
        this.mViewWidth = w;
        this.mViewHeight = h;
        if(this.portrait){
            this.layerSize = h / 4;
            this.layerSpacing = (w - layerSize) / 3;
        }else{
            this.layerSize = w / 4;
            this.layerSpacing = (h - layerSize) / 3;
        }
        this.fieldSize = this.layerSize / 4;
    }

    @Override
    public boolean onTouchEvent(MotionEvent event) {
        if(event.getAction() == MotionEvent.ACTION_DOWN){
            int bX, bY, bZ;
            if(portrait){
                bZ = (int) Math.floor(event.getY() / layerSize);
            }else{
                bZ = (int) Math.floor(event.getX() / layerSize);
            }
            if((portrait && event.getX() >= bZ * layerSpacing && event.getX() < (bZ * layerSpacing) + layerSize)
                    || (!portrait && event.getY() >= bZ * layerSpacing && event.getY() < (bZ * layerSpacing) + layerSize)){
                if(portrait){
                    bY = (int) Math.floor((event.getY() % layerSize) / fieldSize);
                    bX = (int) Math.floor(((event.getX() - (bZ * layerSpacing)) % layerSize) / fieldSize);
                }else{
                    bY = (int) Math.floor(((event.getY() - (bZ * layerSpacing)) % layerSize) / fieldSize);
                    bX = (int) Math.floor((event.getX() % layerSize) / fieldSize);
                }
                if(listener != null && listener.onTap(bX, bY, bZ)){
                    invalidate();
                }
            }
        }
        return true;
    }

}

File: app/src/main/java/lu/kremi151/a3dtictactoe/view/GameBoard.kt
/**
 * 3D Tic Tac Toe for Android
 * Copyright (C) 2018  Michel Kremer (kremi151)
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package lu.kremi151.a3dtictactoe.view

import android.content.Context
import android.graphics.Canvas
import android.graphics.Color
import android.graphics.Paint
import android.util.AttributeSet
import android.view.MotionEvent
import android.view.View
import lu.kremi151.a3dtictactoe.R
import lu.kremi151.a3dtictactoe.enums.FieldValue
import lu.kremi151.a3dtictactoe.interfaces.FieldColorInterceptor
import lu.kremi151.a3dtictactoe.interfaces.OnBoardTapListener
import lu.kremi151.a3dtictactoe.util.GameCube
import kotlin.math.floor

class GameBoard : View {
    private var mContext: Context? = null
    private var fieldPaint: Paint? = null
    private var fieldContentPaint: Paint? = null
    private var fieldValuePaint: Paint? = null
    private var portrait = false
    private var mViewWidth = 0
    private var mViewHeight = 0
    private var layerSize = 0f
    private var layerSpacing = 0f
    private var fieldSize = 0f
    private var listener: OnBoardTapListener? = null
    private var fieldColorInterceptor: FieldColorInterceptor? = null
    private var valueColorInterceptor: FieldColorInterceptor? = null
    private var fieldColorDefault = 0
    private var cube = GameCube()

    constructor(context: Context) : super(context) {
        init(context)
    }

    constructor(context: Context, attrs: AttributeSet?) : super(context, attrs) {
        init(context)
    }

    constructor(context: Context, attrs: AttributeSet?, defStyleAttr: Int) : super(context, attrs, defStyleAttr) {
        init(context)
    }

    constructor(context: Context, attrs: AttributeSet?, defStyleAttr: Int, defStyleRes: Int) : super(context, attrs, defStyleAttr, defStyleRes) {
        init(context)
    }

    fun setCube(cube: GameCube): GameBoard {
        this.cube = cube
        return this
    }

    private fun init(context: Context) {
        mContext = context
        fieldColorDefault = context.resources.getColor(R.color.fieldColor)
        fieldPaint = Paint()
        fieldPaint!!.color = Color.DKGRAY
        fieldPaint!!.strokeWidth = 1f
        fieldPaint!!.style = Paint.Style.STROKE
        fieldValuePaint = Paint(fieldPaint)
        fieldValuePaint!!.strokeWidth = 4f
        fieldValuePaint!!.isAntiAlias = true
        fieldContentPaint = Paint()
        fieldContentPaint!!.color = Color.WHITE
        fieldContentPaint!!.style = Paint.Style.FILL
    }

    fun pause() {}
    fun resume() {}
    fun setListener(listener: OnBoardTapListener?) {
        this.listener = listener
    }

    fun setFieldColorInterceptor(interceptor: FieldColorInterceptor?) {
        fieldColorInterceptor = interceptor
    }

    fun setValueColorInterceptor(interceptor: FieldColorInterceptor?) {
        valueColorInterceptor = interceptor
    }

    override fun onDraw(canvas: Canvas) {
        canvas.save()
        canvas.drawColor(Color.WHITE)
        canvas.translate(1f, 1f)
        for (z in 0 until cube.depth()) {
            for (x in 0 until cube.width()) {
                for (y in 0 until cube.height()) {
                    var left: Float
                    var top: Float
                    if (portrait) {
                        left = z * layerSpacing + x * fieldSize
                        top = z * layerSize + y * fieldSize
                    } else {
                        left = z * layerSize + x * fieldSize
                        top = z * layerSpacing + y * fieldSize
                    }
                    if (fieldColorInterceptor != null) {
                        fieldContentPaint!!.color = fieldColorInterceptor!!.getFieldColor(x, y, z, Color.WHITE)
                        canvas.drawRect(
                                left,
                                top,
                                left + fieldSize,
                                top + fieldSize,
                                fieldContentPaint!!)
                    }
                    canvas.drawRect(
                            left,
                            top,
                            left + fieldSize,
                            top + fieldSize,
                            fieldPaint!!)
                    if (valueColorInterceptor != null) {
                        fieldValuePaint!!.color = valueColorInterceptor!!.getFieldColor(x, y, z, fieldColorDefault)
                    }
                    when (cube.valueAt(x, y, z)) {
                        FieldValue.CIRCLE -> canvas.drawCircle(left + fieldSize / 2, top + fieldSize / 2, fieldSize * 0.4f, fieldValuePaint!!)
                        FieldValue.CROSS -> drawCross(canvas, left + fieldSize * 0.1f, top + fieldSize * 0.1f, left + fieldSize * 0.9f, top + fieldSize * 0.9f, fieldValuePaint)
                    }
                }
            }
        }
        canvas.restore()
    }

    private fun drawCross(canvas: Canvas, left: Float, top: Float, right: Float, bottom: Float, paint: Paint?) {
        canvas.drawLine(left, top, right, bottom, paint!!)
        canvas.drawLine(left, bottom, right, top, paint)
    }

    override fun onSizeChanged(w: Int, h: Int, oldw: Int, oldh: Int) {
        var w = w
        var h = h
        super.onSizeChanged(w, h, oldw, oldh)
        w -= 2
        h -= 2
        portrait = h >= w
        mViewWidth = w
        mViewHeight = h
        if (portrait) {
            layerSize = h / 4.toFloat()
            layerSpacing = (w - layerSize) / 3
        } else {
            layerSize = w / 4.toFloat()
            layerSpacing = (h - layerSize) / 3
        }
        fieldSize = layerSize / 4
    }

    override fun onTouchEvent(event: MotionEvent): Boolean {
        if (event.action == MotionEvent.ACTION_DOWN) {
            val bX: Int
            val bY: Int
            val bZ: Int
            bZ = if (portrait) {
                floor(event.y / layerSize.toDouble()).toInt()
            } else {
                floor(event.x / layerSize.toDouble()).toInt()
            }
            if (portrait && event.x >= bZ * layerSpacing && event.x < bZ * layerSpacing + layerSize
                    || !portrait && event.y >= bZ * layerSpacing && event.y < bZ * layerSpacing + layerSize) {
                if (portrait) {
                    bY = floor(event.y % layerSize / fieldSize.toDouble()).toInt()
                    bX = floor((event.x - bZ * layerSpacing) % layerSize / fieldSize.toDouble()).toInt()
                } else {
                    bY = floor((event.y - bZ * layerSpacing) % layerSize / fieldSize.toDouble()).toInt()
                    bX = floor(event.x % layerSize / fieldSize.toDouble()).toInt()
                }
                if (listener != null && listener!!.onTap(bX, bY, bZ)) {
                    invalidate()
                }
            }
        }
        return true
    }
}